### ISSUE TITLE
iCub SEA legs - Hard and Soft joint limits

### ISSUE THREAD

**Issue by fiorisi** [robotology/QA#246]
Which are the hardware and software joint limits for the iCub SEA legs?

**Comment by fiorisi** [comment]
The issue collects the __HARD__ and the __SOFT__ joint limits for the iCub `SEA legs` (Series Elastic Actuator legs).
The SEA legs are generally assembled on all version of *iCub 2.5* and on the first version of the *Koroibot* robot.  
The joint limits are defined as follows:
* The HARD joint limits are enforced at the *joint level* through a mechanical hard stop (e.g. a pin).
The joint limits have been extracted from the CAD model of the left leg.
The assembly name is: `RC_IIT_018_A_012`.
* The SOFT joint limits are enforced at the *firmware level*.
The values have been copied from the [`iCubGenova04 configuration file`](https://github.com/robotology/robots-configuration/tree/devel/iCubGenova04/hardware/mechanicals).

The following table recaps the joint limits for the complete leg:

|               |    HARD     |     HARD    | SOFT | SOFT |
|---------------|-------------|-------------|-------------|-------------|
| JOINT NAME    | UPPER LIMIT [deg] | LOWER LIMIT [deg] | UPPER LIMIT [deg] | LOWER LIMIT [deg] |
| hip_pitch     |     134     |     -45     | 85 | -30 |
| hip_roll      |     120     |     -20     | 85 | -15 |
| hip_yaw       |      80     |     -80     | 70 | -70 |
| knee          |      4      |     -124    |  0 |-100 |
| ankle_pitch   |      35     |     -35     | -30 | 30 |
| ankle_roll    |      25     |     -25     | 20 | -20 |

## HARD joint limits
### l_hip_pitch
![hippitch](https://user-images.githubusercontent.com/8970230/29408941-72096f40-834a-11e7-9adf-5174a1868839.PNG)
![hippitch2](https://user-images.githubusercontent.com/8970230/29408964-84c2f6e2-834a-11e7-91dc-0ad9ef9fb431.PNG)

### l_hip_roll
![hiproll](https://user-images.githubusercontent.com/8970230/29408982-929b0778-834a-11e7-98a9-9a6148e6680e.PNG)
![hiproll2](https://user-images.githubusercontent.com/8970230/29409003-a712536e-834a-11e7-804c-c8c7017092a7.PNG)

### l_hip_yaw
![hipyaw](https://user-images.githubusercontent.com/8970230/29409017-b7b7346e-834a-11e7-95d7-d80395b618f6.PNG)

### l_knee
![knee](https://user-images.githubusercontent.com/8970230/29409026-c478192a-834a-11e7-9d54-1bd54071192b.PNG)
![knee2](https://user-images.githubusercontent.com/8970230/29409031-c8dae95c-834a-11e7-8b11-4658578162b4.PNG)

### l_ankle_pitch
![anklepitch](https://user-images.githubusercontent.com/8970230/29409037-d12ec1a0-834a-11e7-98a2-e3bc1a9cda71.PNG)
![anklepitch2](https://user-images.githubusercontent.com/8970230/29409039-d4b5832c-834a-11e7-88b8-736f363656de.PNG)

### l_ankle_roll
![ankleroll](https://user-images.githubusercontent.com/8970230/29409055-deb8c8f2-834a-11e7-90c1-156b3ce3c19e.PNG)

**Comment by fiorisi** [comment]
cc: @traversaro @DanielePucci @francesco-romano @gabrielenava @S-Dafarra @pattacini 
related issues: 
* https://github.com/loc2/element_icub-at-ihmc/issues/2
* https://github.com/robotology-playground/icub-model-generator/issues/48
* https://github.com/robotology-playground/icub-model-generator/issues/5

**Comment by DanielePucci** [comment]
Hi @fiorisi, thanks a lot for the effort. Crystal.

Did you by chance check that these limits are met also on a "real" robot we have in the lab?

**Comment by fiorisi** [comment]
The joint limits have been checked on the real robot. The problem is that in this case it is necessary to take into account cover self-collisions.
The joint limits have been measured with the Yarp robot motor GUI from the left leg of the `iCubGenova04` robot.

The following table recaps the joint limits for the complete leg:

|               | REAL ROBOT | REAL ROBOT |
|---------------|-------------|-------------|
| JOINT NAME    | UPPER LIMIT [deg] | LOWER LIMIT [deg] |
| hip_pitch     |     100     |     -20 \*     |
| hip_roll      |      85     |      -10 \*    |
| hip_yaw       |     55 \*    |       -55 \*   |
| knee          |      4     |     -110    |
| ankle_pitch   |     34     |      -34 \**    |
| ankle_roll    |     25     |      -25    |

__(\*) NOTE__: the joint limits of the hip have complex interdependencies. If the SOFT limits are used, when the joint *hip_yaw* is rotated close to its limit (either 70 or -70), the covers can collide with the waist when moving the other hip joints (namely, *hip_roll* and *hip_pitch*). For this reason the REAL ROBOT limits have a smaller range of motion.

__(\**) NOTE__: in the CAD model, the joint has a cover collision before reaching the limit value. Nevertheless, probably due to deformations during the production of the covers, there is no collision on the iCubGenova04.  
This problem will be fixed ASAP in the CAD model.

### l_hip_pitch
![hippitch_real](https://user-images.githubusercontent.com/8970230/29412948-05a3ed06-835b-11e7-8e46-ed5137d5aafb.JPG)
![hippitch_real2](https://user-images.githubusercontent.com/8970230/29412964-142c5980-835b-11e7-84fe-6c44e532009e.JPG)

### l_hip_roll
![hiproll_real](https://user-images.githubusercontent.com/8970230/29412984-2f5e7332-835b-11e7-88ba-d42b36b2a4c0.JPG)
![hiproll_real2](https://user-images.githubusercontent.com/8970230/29412991-34fe8f16-835b-11e7-90bb-4a01ec5ceb71.JPG)

### l_hip_yaw
![hipyaw_real](https://user-images.githubusercontent.com/8970230/29413016-3ecaa746-835b-11e7-8ab6-4fed576ce329.JPG)
![hipyaw_real2](https://user-images.githubusercontent.com/8970230/29413023-431f9888-835b-11e7-97f8-a910bfedb493.JPG)

### l_knee
![knee_real](https://user-images.githubusercontent.com/8970230/29413041-5531611e-835b-11e7-910c-46374a1d7c9b.JPG)
![knee_real2](https://user-images.githubusercontent.com/8970230/29413045-58e97fd0-835b-11e7-9e7d-a0325171be94.JPG)

### l_ankle_pitch
![anklepitch_real](https://user-images.githubusercontent.com/8970230/29413054-64c05914-835b-11e7-9947-9c636fa333ac.JPG)
![anklepitch_real2](https://user-images.githubusercontent.com/8970230/29413061-692b653e-835b-11e7-88ee-03c500a41cb0.JPG)

### l_ankle_roll
![ankleroll_real](https://user-images.githubusercontent.com/8970230/29413081-72d5f13a-835b-11e7-8fc8-a84269202cf2.JPG)
![ankleroll_real2](https://user-images.githubusercontent.com/8970230/29413084-78216b10-835b-11e7-8c69-0cab1763aef3.JPG)

**Comment by DanielePucci** [comment]
Nice. I think we can also update the corresponding urdf models then

@traversaro @francesco-romano @pattacini

**Comment by traversaro** [comment]
The file to update is https://github.com/robotology-playground/icub-model-generator/blob/master/simmechanics/data/icub2_5/ICUB_2-5_BB_joint_parameters.csv . 

We should also update the iCubHeidelberg01 model, but that one is still not integrated in the repo for some issues: 
* https://github.com/robotology-playground/icub-model-generator/pull/15
* https://github.com/robotology-playground/icub-model-generator/issues/35 
* https://github.com/robotology-playground/icub-model-generator/issues/16

**Comment by serena-ivaldi** [comment]
That applies also to our iCubNancy01, right? In that case we should update also our URDF, right? Can we assume the same values for the covers (collisions)?

@dogoepp @InriaBrice @olivierrochel-inria

**Comment by pattacini** [comment]
@fiorisi could you take on @serena-ivaldi's question?

**Comment by randaz81** [comment]
Please open a new issue for your robot

**Comment by S-Dafarra** [comment]
After a double check on the robot with @fiorisi we noticed that the limits were bigger than what written initially. The table in the second comment has been updated together with the corresponding CAD images. This time the range has been measured by constraining the pin to be coincident with the hard limit surface (initially the range of motion was measured considering only the hollow, but the pin as a smaller radius, causing the inconsistency noticed above). 

Notice that it is also possible to reach all the leg joints hard limits, except from the knee lower limit due to the cabling. All the other limits are more or less trivially reachable even with the covers mounted. The tricky ones to be reached are the hip roll limits. The upper limit can be reached by bending the leg in this pretty awkward position, where also the pitch and the yaw are close to their limits:
![photo_2018-04-12_16-55-16](https://user-images.githubusercontent.com/18591940/38685959-7ee97a0e-3e73-11e8-8d20-bd8c27370425.jpg)


The hip roll lower limit instead can be approached when the robot is **not** on the pole, even if it cannot be mechanically reached because the upper leg is hitting the hip_pitch pulley:
![photo_2018-04-12_17-11-53](https://user-images.githubusercontent.com/18591940/38686713-463b3b14-3e75-11e8-9bf9-4cd61a53cd4b.jpg)
![photo_2018-04-12_17-27-21](https://user-images.githubusercontent.com/18591940/38687369-d2d756e2-3e76-11e8-8d19-96c11b7be4c7.jpg)
